File: agents/github-action-agent/src/tonic_agent/hydrate.py
"""Fetch base/head file text from GitHub for PR analysis (compare API + contents)."""

from __future__ import annotations

import base64
import os
from typing import Any

from urllib.parse import quote

from .github_api import fetch_url_text_authenticated, get_git_blob_text, get_json


def _api_base() -> str:
    return os.environ.get("GITHUB_API_URL", "https://api.github.com").rstrip("/")


BINARY_EXTENSIONS = frozenset(
    {
        ".png",
        ".jpg",
        ".jpeg",
        ".gif",
        ".webp",
        ".ico",
        ".pdf",
        ".zip",
        ".gz",
        ".tgz",
        ".bz2",
        ".7z",
        ".rar",
        ".exe",
        ".dll",
        ".so",
        ".dylib",
        ".woff",
        ".woff2",
        ".ttf",
        ".eot",
        ".mp3",
        ".mp4",
        ".webm",
        ".wasm",
        ".pyc",
        ".class",
        ".jar",
    }
)


def is_probably_text_path(path: str) -> bool:
    lower = path.lower()
    for ext in BINARY_EXTENSIONS:
        if lower.endswith(ext):
            return False
    return True


def _encode_path(path: str) -> str:
    return "/".join(quote(segment, safe="") for segment in path.split("/"))


def compare_commits(owner: str, repo: str, base_sha: str, head_sha: str, token: str) -> dict[str, Any]:
    url = f"{_api_base()}/repos/{owner}/{repo}/compare/{base_sha}...{head_sha}"
    return get_json(url, token)  # type: ignore[return-value]


def get_contents_payload(
    owner: str, repo: str, path: str, ref: str, token: str
) -> dict[str, Any] | None:
    url = f"{_api_base()}/repos/{owner}/{repo}/contents/{_encode_path(path)}?ref={ref}"
    try:
        return get_json(url, token)  # type: ignore[assignment]
    except RuntimeError as e:
        if "GitHub API 404" in str(e):
            return None
        raise


def decode_file_content(payload: dict[str, Any]) -> str | None:
    enc = payload.get("encoding")
    content = payload.get("content")
    if enc == "base64" and isinstance(content, str):
        raw = base64.b64decode(content)
        if b"\x00" in raw:
            return None
        try:
            return raw.decode("utf-8")
        except UnicodeDecodeError:
            return raw.decode("utf-8", errors="replace")
    return None


def fetch_file_text(
    owner: str,
    repo: str,
    path: str,
    ref: str,
    token: str,
    *,
    blob_sha: str | None = None,
) -> str | None:
    """
    Prefer Contents API; for large files use download_url; optional git blob when sha known
    (compare.files[].sha is head blob for that path).
    """
    payload = get_contents_payload(owner, repo, path, ref, token)
    if not payload or payload.get("type") != "file":
        if blob_sha:
            return get_git_blob_text(owner, repo, blob_sha, token)
        return None
    text = decode_file_content(payload)
    if text is None and isinstance(payload.get("download_url"), str):
        du = str(payload["download_url"])
        text = fetch_url_text_authenticated(du, token)
    if text is None and blob_sha:
        text = get_git_blob_text(owner, repo, blob_sha, token)
    return text


def list_tree_blob_shas(owner: str, repo: str, tree_sha: str, token: str) -> dict[str, str]:
    url = f"{_api_base()}/repos/{owner}/{repo}/git/trees/{tree_sha}?recursive=1"
    data = get_json(url, token)
    out: dict[str, str] = {}
    if not isinstance(data, dict):
        return out
    for item in data.get("tree") or []:
        if not isinstance(item, dict):
            continue
        if item.get("type") != "blob":
            continue
        p = item.get("path")
        s = item.get("sha")
        if isinstance(p, str) and isinstance(s, str):
            out[p] = s
    return out


def _lines(text: str | None) -> list[str]:
    if not text:
        return []
    return text.splitlines()


def hydrate_pr_files(
    owner: str,
    repo: str,
    base_sha: str,
    head_sha: str,
    token: str,
    *,
    mode: str,
    max_files: int,
) -> dict[str, tuple[list[str], list[str], str]]:
    """Return path -> (base lines, head lines, git status hint)."""
    out: dict[str, tuple[list[str], list[str], str]] = {}
    if mode == "symmetric-union":
        base_tree = list_tree_blob_shas(owner, repo, base_sha, token)
        head_tree = list_tree_blob_shas(owner, repo, head_sha, token)
        paths = sorted(set(base_tree) | set(head_tree))
        for path in paths:
            if len(out) >= max_files:
                break
            if not is_probably_text_path(path):
                continue
            bs = base_tree.get(path)
            hs = head_tree.get(path)
            if bs is not None and bs == hs:
                continue
            left_t = (
                fetch_file_text(owner, repo, path, base_sha, token, blob_sha=bs)
                if path in base_tree
                else None
            )
            right_t = (
                fetch_file_text(owner, repo, path, head_sha, token, blob_sha=hs)
                if path in head_tree
                else None
            )
            if left_t is None and bs and path in base_tree:
                left_t = get_git_blob_text(owner, repo, bs, token)
            if right_t is None and hs and path in head_tree:
                right_t = get_git_blob_text(owner, repo, hs, token)
            if path in base_tree and path not in head_tree:
                st = "removed"
            elif path not in base_tree and path in head_tree:
                st = "added"
            else:
                st = "modified"
            out[path] = (_lines(left_t), _lines(right_t), st)
        return out

    cmp = compare_commits(owner, repo, base_sha, head_sha, token)
    files = cmp.get("files") or []
    if not isinstance(files, list):
        return out
    for ent in files:
        if len(out) >= max_files:
            break
        if not isinstance(ent, dict):
            continue
        status = str(ent.get("status") or "")
        filename = ent.get("filename")
        if not isinstance(filename, str):
            continue
        if not is_probably_text_path(filename):
            continue
        prev = ent.get("previous_filename")
        prev_s = prev if isinstance(prev, str) else None
        head_blob_sha = ent.get("sha") if isinstance(ent.get("sha"), str) else None

        left_t: str | None
        right_t: str | None
        if status == "added":
            left_t = None
            right_t = fetch_file_text(
                owner, repo, filename, head_sha, token, blob_sha=head_blob_sha
            )
        elif status == "removed":
            left_t = fetch_file_text(owner, repo, filename, base_sha, token)
            right_t = None
        elif status == "renamed" and prev_s:
            left_t = fetch_file_text(owner, repo, prev_s, base_sha, token)
            right_t = fetch_file_text(
                owner, repo, filename, head_sha, token, blob_sha=head_blob_sha
            )
        else:
            left_t = fetch_file_text(owner, repo, filename, base_sha, token)
            right_t = fetch_file_text(
                owner, repo, filename, head_sha, token, blob_sha=head_blob_sha
            )
        out[filename] = (_lines(left_t), _lines(right_t), status or "modified")
    return out

File: agents/github-action-agent-node/src/hydrate.ts
import { fetchUrlTextAuthenticated, getGitBlobText, githubGetJson } from "./githubApi";
import type { GitMergeHydrationOptions } from "@mergetonic/core";
import { hydrateGitMerge, type GitHydratePair } from "./hydrateGitMerge";

const BINARY = new Set([
  ".png",
  ".jpg",
  ".jpeg",
  ".gif",
  ".webp",
  ".ico",
  ".pdf",
  ".zip",
  ".gz",
  ".tgz",
  ".bz2",
  ".7z",
  ".rar",
  ".exe",
  ".dll",
  ".so",
  ".dylib",
  ".woff",
  ".woff2",
  ".ttf",
  ".eot",
  ".mp3",
  ".mp4",
  ".webm",
  ".wasm",
  ".pyc",
  ".class",
  ".jar",
]);

export function isProbablyTextPath(path: string): boolean {
  const lower = path.toLowerCase();
  for (const ext of BINARY) {
    if (lower.endsWith(ext)) {
      return false;
    }
  }
  return true;
}

function encodePath(path: string): string {
  return path
    .split("/")
    .map((s) => encodeURIComponent(s))
    .join("/");
}

export async function compareCommits(
  owner: string,
  repo: string,
  baseSha: string,
  headSha: string,
  token: string,
): Promise<Record<string, unknown>> {
  const api = (process.env.GITHUB_API_URL ?? "https://api.github.com").replace(/\/$/, "");
  const url = `${api}/repos/${owner}/${repo}/compare/${baseSha}...${headSha}`;
  return (await githubGetJson(url, token)) as Record<string, unknown>;
}

export async function getContentsPayload(
  owner: string,
  repo: string,
  path: string,
  ref: string,
  token: string,
): Promise<Record<string, unknown> | null> {
  const api = (process.env.GITHUB_API_URL ?? "https://api.github.com").replace(/\/$/, "");
  const url = `${api}/repos/${owner}/${repo}/contents/${encodePath(path)}?ref=${encodeURIComponent(ref)}`;
  try {
    return (await githubGetJson(url, token)) as Record<string, unknown>;
  } catch (e) {
    if (e instanceof Error && e.message.includes("GitHub API 404")) {
      return null;
    }
    throw e;
  }
}

export function decodeFileContent(payload: Record<string, unknown>): string | null {
  const enc = payload.encoding;
  const content = payload.content;
  if (enc === "base64" && typeof content === "string") {
    const raw = Buffer.from(content, "base64");
    if (raw.includes(0)) {
      return null;
    }
    return raw.toString("utf8");
  }
  return null;
}

export async function fetchFileText(
  owner: string,
  repo: string,
  path: string,
  ref: string,
  token: string,
  blobSha?: string | null,
): Promise<string | null> {
  const payload = await getContentsPayload(owner, repo, path, ref, token);
  if (!payload || payload.type !== "file") {
    if (blobSha) {
      return getGitBlobText(owner, repo, blobSha, token);
    }
    return null;
  }
  let text = decodeFileContent(payload);
  if (text == null && typeof payload.download_url === "string") {
    text = await fetchUrlTextAuthenticated(payload.download_url, token);
  }
  if (text == null && blobSha) {
    text = await getGitBlobText(owner, repo, blobSha, token);
  }
  return text;
}

export async function listTreeBlobShas(
  owner: string,
  repo: string,
  treeSha: string,
  token: string,
): Promise<Record<string, string>> {
  const api = (process.env.GITHUB_API_URL ?? "https://api.github.com").replace(/\/$/, "");
  const url = `${api}/repos/${owner}/${repo}/git/trees/${treeSha}?recursive=1`;
  const data = (await githubGetJson(url, token)) as Record<string, unknown>;
  const out: Record<string, string> = {};
  const tree = data.tree;
  if (!Array.isArray(tree)) {
    return out;
  }
  for (const item of tree) {
    const ent = item as Record<string, unknown>;
    if (ent.type !== "blob") {
      continue;
    }
    const p = ent.path;
    const s = ent.sha;
    if (typeof p === "string" && typeof s === "string") {
      out[p] = s;
    }
  }
  return out;
}

function lines(text: string | null | undefined): string[] {
  if (!text) {
    return [];
  }
  const out = text.split(/\r?\n/);
  if (out.length > 0 && out[out.length - 1] === "") {
    out.pop();
  }
  return out;
}

export type HydratePair = {
  leftLines: string[];
  rightLines: string[];
  status: string;
  gitAnnotatedLines?: string[];
  mergedLines?: string[];
};

export async function hydratePrFiles(
  owner: string,
  repo: string,
  baseSha: string,
  headSha: string,
  token: string,
  mode: string,
  maxFiles: number,
  gitMergeHydration?: GitMergeHydrationOptions,
): Promise<Record<string, HydratePair>> {
  if (mode === "git-merge") {
    const workspace = process.env.TONIC_AGENT_ISOLATED_WORKSPACE || process.env.GITHUB_WORKSPACE || process.cwd();
    return hydrateGitMerge({
      workspace,
      baseSha,
      headSha,
      maxFiles,
      gitMergeHydration,
    }) as Record<string, GitHydratePair>;
  }
  const out: Record<string, HydratePair> = {};
  if (mode === "symmetric-union") {
    const baseTree = await listTreeBlobShas(owner, repo, baseSha, token);
    const headTree = await listTreeBlobShas(owner, repo, headSha, token);
    const paths = Array.from(new Set([...Object.keys(baseTree), ...Object.keys(headTree)])).sort();
    for (const path of paths) {
      if (Object.keys(out).length >= maxFiles) {
        break;
      }
      if (!isProbablyTextPath(path)) {
        continue;
      }
      const bs = baseTree[path];
      const hs = headTree[path];
      if (bs != null && bs === hs) {
        continue;
      }
      let leftT =
        path in baseTree
          ? await fetchFileText(owner, repo, path, baseSha, token, bs ?? null)
          : null;
      let rightT =
        path in headTree
          ? await fetchFileText(owner, repo, path, headSha, token, hs ?? null)
          : null;
      if (leftT == null && bs && path in baseTree) {
        leftT = await getGitBlobText(owner, repo, bs, token);
      }
      if (rightT == null && hs && path in headTree) {
        rightT = await getGitBlobText(owner, repo, hs, token);
      }
      let st: string;
      if (path in baseTree && !(path in headTree)) {
        st = "removed";
      } else if (!(path in baseTree) && path in headTree) {
        st = "added";
      } else {
        st = "modified";
      }
      out[path] = { leftLines: lines(leftT), rightLines: lines(rightT), status: st };
    }
    return out;
  }

  const cmp = await compareCommits(owner, repo, baseSha, headSha, token);
  const files = cmp.files;
  if (!Array.isArray(files)) {
    return out;
  }
  for (const ent of files) {
    if (Object.keys(out).length >= maxFiles) {
      break;
    }
    const e = ent as Record<string, unknown>;
    const status = String(e.status ?? "");
    const filename = e.filename;
    if (typeof filename !== "string" || !isProbablyTextPath(filename)) {
      continue;
    }
    const prev = e.previous_filename;
    const prevS = typeof prev === "string" ? prev : null;
    const headBlobSha = typeof e.sha === "string" ? e.sha : null;

    let leftT: string | null;
    let rightT: string | null;
    if (status === "added") {
      leftT = null;
      rightT = await fetchFileText(owner, repo, filename, headSha, token, headBlobSha);
    } else if (status === "removed") {
      leftT = await fetchFileText(owner, repo, filename, baseSha, token);
      rightT = null;
    } else if (status === "renamed" && prevS) {
      leftT = await fetchFileText(owner, repo, prevS, baseSha, token);
      rightT = await fetchFileText(owner, repo, filename, headSha, token, headBlobSha);
    } else {
      leftT = await fetchFileText(owner, repo, filename, baseSha, token);
      rightT = await fetchFileText(owner, repo, filename, headSha, token, headBlobSha);
    }
    out[filename] = {
      leftLines: lines(leftT),
      rightLines: lines(rightT),
      status: status || "modified",
    };
  }
  return out;
}
